From 180393d6916538a2cbe076021fa7fda79420ea7d Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:01:04 -0400
Subject: [infiniband] mlx5: Add MR to radix tree in reg_mr_callback

Message-id: <75c91194360743c29c2847f68cf523b4171b1b6d.1410425016.git.avadai@redhat.com>
Patchwork-id: 93649
O-Subject: [Patch RHEL7.1 V1 128/197] IB/mlx5: Add MR to radix tree in reg_mr_callback
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Haggai Eran <haggaie@mellanox.com>

For memory regions that are allocated using reg_umr, the suffix of
mlx5_core_create_mkey isn't being called.  Instead the creation is
completed in a callback function (reg_mr_callback).  This means that
these MRs aren't being added to the MR radix tree.  Add them in the
callback.

Signed-off-by: Haggai Eran <haggaie@mellanox.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 8605933a22796243982e7ed838deca5549c64c62)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/mlx5/mr.c b/drivers/infiniband/hw/mlx5/mr.c
index ad58985..9d932a2 100644
--- a/drivers/infiniband/hw/mlx5/mr.c
+++ b/drivers/infiniband/hw/mlx5/mr.c
@@ -73,6 +73,8 @@ static void reg_mr_callback(int status, void *context)
  struct mlx5_cache_ent *ent = &cache->ent[c];
  u8 key;
  unsigned long flags;
+ struct mlx5_mr_table *table = &dev->mdev.priv.mr_table;
+ int err;
 
  spin_lock_irqsave(&ent->lock, flags);
  ent->pending--;
@@ -107,6 +109,13 @@ static void reg_mr_callback(int status, void *context)
  ent->cur++;
  ent->size++;
  spin_unlock_irqrestore(&ent->lock, flags);
+
+ write_lock_irqsave(&table->lock, flags);
+ err = radix_tree_insert(&table->tree, mlx5_base_mkey(mr->mmr.key),
+    &mr->mmr);
+ if (err)
+  pr_err("Error inserting to mr tree. 0x%x\n", -err);
+ write_unlock_irqrestore(&table->lock, flags);
 }
 
 static int add_keys(struct mlx5_ib_dev *dev, int c, int num)
-- 
1.7.1